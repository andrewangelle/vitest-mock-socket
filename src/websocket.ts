import type { Client, CloseOptions } from 'mock-socket';
import { Server } from 'mock-socket';

import type { DeserializedMessage, WebSocketServerOptions } from './types';

export class WebSocketServer {
  static clients: WebSocketServer[] = [];

  static clean() {
    for (const client of WebSocketServer.clients) {
      client.close();
      client.messages = [];
    }
    WebSocketServer.clients = [];
  }

  server: Server;
  messages: DeserializedMessage[] = [];
  #options: WebSocketServerOptions;
  #isConnected = Promise.withResolvers<Client>();
  #isClosed = Promise.withResolvers<Client>();
  #messageAdded = Promise.withResolvers<void>();
  #messageQueue: DeserializedMessage[] = [];

  constructor(
    url: string,
    options: WebSocketServerOptions = { jsonProtocol: false },
  ) {
    WebSocketServer.clients.push(this);
    this.#options = options;
    this.server = this.#createServer(url);
  }

  async connected() {
    return this.#isConnected.promise;
  }

  async closed() {
    return this.#isClosed.promise;
  }

  async nextMessage() {
    // if we have a queued message return it immediately
    if (this.#messageQueue.length > 0) {
      return this.#messageQueue.shift();
    }

    // otherwise wait until we have one
    await this.#messageAdded.promise;
    return this.#messageQueue.shift();
  }

  on(
    eventName: 'connection' | 'message' | 'close',
    callback: (socket: Client) => void,
  ): void {
    this.server.on(eventName, callback);
  }

  send(message: DeserializedMessage) {
    this.server.emit('message', this.#serializer(message));
  }

  close(options?: CloseOptions) {
    this.server.close(options);
  }

  error(options?: CloseOptions) {
    this.server.emit('error', null);
    this.server.close(options);
  }

  #serializer(deserializedMessage: DeserializedMessage) {
    if (this.#options.jsonProtocol) {
      return JSON.stringify(deserializedMessage);
    }
    return deserializedMessage;
  }

  #deserializer(serializedMessage: string): DeserializedMessage {
    if (this.#options.jsonProtocol) {
      try {
        const safeParsed = JSON.parse(serializedMessage);
        return safeParsed;
      } catch {
        // silently swallow JSON parse error
      }
    }
    return serializedMessage;
  }

  #createServer(url: string) {
    const server = new Server(url, this.#options);

    server.on('close', (socket: Client) => {
      this.#isClosed.resolve(socket);
      this.#isClosed = Promise.withResolvers();
    });

    server.on('connection', (socket: Client) => {
      this.#isConnected.resolve(socket);
      this.#isConnected = Promise.withResolvers();

      socket.on('message', (message) => {
        const parsedMessage = this.#deserializer(message as string);
        this.messages.push(parsedMessage);
        this.#messageQueue.push(parsedMessage);
        this.#messageAdded.resolve();
        this.#messageAdded = Promise.withResolvers();
      });
    });

    return server;
  }
}
