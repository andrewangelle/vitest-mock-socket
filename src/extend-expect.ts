import type { MatcherState } from '@vitest/expect';
import { diff } from '@vitest/utils/diff';
import { expect } from 'vitest';

import type { DeserializedMessage, ReceiveMessageOptions } from './types';
import {
  createFmt,
  getInvalidServerResult,
  getNextMessageOrTimeout,
  getTimedOutResult,
  isTimeout,
} from './utils';
import { WebSocketServer } from './websocket';

interface CustomMatchers<MatcherResult = unknown> {
  toReceiveMessage<MessageType = object>(
    message: DeserializedMessage<MessageType>,
    options?: ReceiveMessageOptions,
  ): Promise<MatcherResult>;
  toHaveReceivedMessages<MessageType = object>(
    messages: DeserializedMessage<MessageType>[],
  ): MatcherResult;
}

declare module '@vitest/expect' {
  // biome-ignore lint/suspicious/noExplicitAny: required to extend vitest matchers
  interface Assertion<T = any> extends CustomMatchers<T> {}
  interface AsymmetricMatchersContaining extends CustomMatchers {}
}

expect.extend({
  /**
   *
   * @name toHaveReceivedMessages
   * @description
   *
   * 1. Validate that a websocket server was passed to expect.
   *    i.e. expect(server).toHaveReceivedMessages(...)
   *
   * 2. Test and print the results
   */
  toHaveReceivedMessages(
    this: MatcherState,
    received: DeserializedMessage[],
    expected: DeserializedMessage[],
  ) {
    if (!(received instanceof WebSocketServer)) {
      return getInvalidServerResult.call(
        this,
        'toHaveReceivedMessages',
        received,
      );
    }

    // Handle option.jsonProtocol set to true
    const equalities = expected.map((expectedMsg) =>
      received.messages.some((receivedMsg) =>
        this.equals(receivedMsg, expectedMsg),
      ),
    );

    const pass = this.isNot
      ? equalities.some(Boolean)
      : equalities.every(Boolean);

    return {
      actual: received.messages,
      expected,
      pass,
      message: () => {
        const printCli = createFmt.call(this, 'toHaveReceivedMessages');

        if (this.isNot) {
          return printCli`
            Expected the WebSocketServer to not have received the following messages: 
              ${this.utils.printExpected(expected)}
            
            But it received: 
              ${this.utils.printReceived(received.messages)}
          `;
        }

        return printCli`
          Expected the WebSocketServer to have received the following messages: 
            ${this.utils.printExpected(expected)}

          Received: 
            ${this.utils.printReceived(received.messages)}
        `;
      },
    };
  },

  /**
   *
   * @name toReceiveMessage
   * @description
   *
   * 1. Validate that a websocket server was passed to expect.
   *    i.e. await expect(server).toReceiveMessage(...)
   *
   * 2. Wait for the message or timeout. Fail if we timeout first.
   *
   * 3. Test and print the results
   *
   */
  async toReceiveMessage(
    this: MatcherState,
    received: DeserializedMessage,
    expected: DeserializedMessage,
    options?: ReceiveMessageOptions,
  ) {
    const printCli = createFmt.call(this, 'toReceiveMessage');

    if (!(received instanceof WebSocketServer)) {
      return getInvalidServerResult.call(this, 'toReceiveMessage', received);
    }

    const result = await getNextMessageOrTimeout(received, options);

    if (isTimeout(result)) {
      return getTimedOutResult.call(this, options);
    }

    const pass = this.equals(result, expected);

    return {
      actual: result,
      expected,
      pass,
      message: () => {
        if (pass) {
          return printCli`
            Expected the next received message to not equal:
              ${this.utils.printExpected(expected)}

            Received:
              ${this.utils.printReceived(result)}
          `;
        }

        return printCli`
          Expected the next received message to equal:
            ${this.utils.printExpected(expected)}

          Received:
            ${this.utils.printReceived(result)}

          Difference:
            ${diff(expected, result, { expand: this.expand }) ?? ''}
        `;
      },
    };
  },
});
