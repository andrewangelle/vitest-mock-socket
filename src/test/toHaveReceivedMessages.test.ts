import '../extend-expect';

import { afterEach, beforeEach, describe, expect, it } from 'vitest';

import { WebSocketServer } from '../websocket';

let server: WebSocketServer;
let client: WebSocket;
beforeEach(async () => {
  server = new WebSocketServer('ws://localhost:1234');
  client = new WebSocket('ws://localhost:1234');
  await server.connected();
});

afterEach(() => {
  WebSocketServer.clean();
});

describe('.toHaveReceivedMessages', () => {
  it('passes when the websocket server received the expected messages', async () => {
    client.send('hello there');
    client.send('how are you?');
    client.send('good?');
    await server.nextMessage();
    await server.nextMessage();
    await server.nextMessage();
    expect(server).toHaveReceivedMessages(['hello there', 'good?']);
  });

  it('passes when the websocket server received the expected JSON messages', async () => {
    const jsonServer = new WebSocketServer('ws://localhost:9876', {
      jsonProtocol: true,
    });
    const jsonClient = new WebSocket('ws://localhost:9876');
    await jsonServer.connected();
    jsonClient.send(`{"type":"GREETING","payload":"hello there"}`);
    jsonClient.send(`{"type":"GREETING","payload":"how are you?"}`);
    jsonClient.send(`{"type":"GREETING","payload":"good?"}`);
    await jsonServer.nextMessage();
    await jsonServer.nextMessage();
    await jsonServer.nextMessage();
    expect(jsonServer).toHaveReceivedMessages([
      { type: 'GREETING', payload: 'good?' },
      { type: 'GREETING', payload: 'hello there' },
    ]);
  });

  // TODO: Fix Array indentation
  it('fails when the websocket server did not receive the expected messages', async () => {
    client.send('hello there');
    client.send('how are you?');
    client.send('good?');
    await server.nextMessage();
    await server.nextMessage();
    await server.nextMessage();
    expect(() => {
      expect(server).toHaveReceivedMessages(['hello there', "'sup?"]);
    }).toThrowErrorMatchingSnapshot();
  });

  it('fails when called with an expected argument that is not a valid WS', async () => {
    expect(() => {
      expect('boom').toHaveReceivedMessages(['hello there']);
    }).toThrowErrorMatchingSnapshot();
  });
});

describe('.not.toHaveReceivedMessages', () => {
  it('passes when the websocket server received none of the specified messages', async () => {
    client.send('hello there');
    client.send('how are you?');
    client.send('good?');
    await server.nextMessage();
    await server.nextMessage();
    await server.nextMessage();
    expect(server).not.toHaveReceivedMessages(["'sup?", 'U good?']);
  });

  it('fails when the websocket server received at least one unexpected message', async () => {
    client.send('hello there');
    client.send('how are you?');
    client.send('good?');
    await server.nextMessage();
    await server.nextMessage();
    await server.nextMessage();
    expect(() => {
      expect(server).not.toHaveReceivedMessages([
        "'sup?",
        'U good?',
        'hello there',
      ]);
    }).toThrowErrorMatchingSnapshot();
  });

  it('fails when called with an expected argument that is not a valid WS', async () => {
    expect(() => {
      expect('boom').not.toHaveReceivedMessages(['hello there']);
    }).toThrowErrorMatchingSnapshot();
  });
});
