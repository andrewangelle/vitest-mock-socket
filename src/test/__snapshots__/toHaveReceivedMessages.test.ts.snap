// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`.not.toHaveReceivedMessages > fails when called with an expected argument that is not a valid WS 1`] = `
[Error: [2mexpect([22m[31mWebSocketServer[39m[2m).not.toHaveReceivedMessages([22m[32mexpected[39m[2m)[22m

Expected the websocket object to be a valid WebSocketServer mock.
Received: string
  [31m"boom"[39m]
`;

exports[`.not.toHaveReceivedMessages > fails when the websocket server received at least one unexpected message 1`] = `
[Error: [2mexpect([22m[31mWebSocketServer[39m[2m).not.toHaveReceivedMessages([22m[32mexpected[39m[2m)[22m

Expected the WebSocketServer to not have received the following messages:

  [32mArray [
    "'sup?",
    "U good?",
    "hello there",
  ][39m

But it received:

  [31mArray [
    "hello there",
    "how are you?",
    "good?",
  ][39m
]
`;

exports[`.toHaveReceivedMessages > fails when called with an expected argument that is not a valid WS 1`] = `
[Error: [2mexpect([22m[31mWebSocketServer[39m[2m).toHaveReceivedMessages([22m[32mexpected[39m[2m)[22m

Expected the websocket object to be a valid WebSocketServer mock.
Received: string
  [31m"boom"[39m]
`;

exports[`.toHaveReceivedMessages > fails when the websocket server did not receive the expected messages 1`] = `
[Error: [2mexpect([22m[31mWebSocketServer[39m[2m).toHaveReceivedMessages([22m[32mexpected[39m[2m)[22m

Expected the WebSocketServer to have received the following messages:

  [32mArray [
    "hello there",
    "'sup?",
  ][39m

Received:

  [31mArray [
    "hello there",
    "how are you?",
    "good?",
  ][39m
]
`;
